File: SVM.r
##################################### problem1 ##################################################
#####Support Vector Machines 
library(readr)
# Load the Dataset
Salary_train <- read.csv("E:\\DATA SCIENCE ASSIGNMENT\\Class And Assignment Dataset\\Asss\\Black  Box Tech (Support Vector Machine)\\SalaryData_Train.csv", stringsAsFactors = TRUE)
Salary_test <- read.csv("E:\\DATA SCIENCE ASSIGNMENT\\Class And Assignment Dataset\\Asss\\Black  Box Tech (Support Vector Machine)\\SalaryData_Test.csv",stringsAsFactors = TRUE)

summary(Salary_train)
summary(Salary_test)
str(Salary_test)
str(Salary_train)

#checking column names
colnames(Salary_train)
colnames(Salary_test)

# Training a model on the data ----
# Begin by training a simple linear SVM
install.packages("kernlab")
library(kernlab)

salary_classifier <- ksvm(Salary ~ ., data = Salary_train, kernel = "vanilladot")
?ksvm

## Evaluating model performance ----
# predictions on testing dataset
salary_predictions <- predict(salary_classifier, Salary_test)

table(salary_predictions, Salary_test$Salary)
agreement <- salary_predictions == Salary_test$Salary
table(agreement)
prop.table(table(agreement))

## Improving model performance ----
salary_classifier_rbf <- ksvm(Salary ~ ., data = Salary_train, kernel = "rbfdot")
salary_predictions_rbf <- predict(salary_classifier_rbf, Salary_test)
agreement_rbf <- salary_predictions_rbf == Salary_test$Salary
table(agreement_rbf)
prop.table(table(agreement_rbf))

#85% the people who are living in that area having good salaries.Built properties can be easily sell.

########################### problem2 #################################################################
#####Support Vector Machines 

# Load the Dataset
forest_fires_data <- read.csv("E:\\DATA SCIENCE ASSIGNMENT\\Class And Assignment Dataset\\Asss\\Black  Box Tech (Support Vector Machine)\\forestfires.csv", stringsAsFactors = TRUE)

summary(forest_fires_data)
colnames(forest_fires_data)

# Partition Data into train and test data
forest_train <- forest_fires_data[1:414, ]
forest_test  <- forest_fires_data[415:517, ]

# Training a model on the data ----
# Begin by training a simple linear SVM
install.packages("kernlab")
library(kernlab)

forest_classifier <- ksvm(size_category ~ ., data = forest_train, kernel = "vanilladot")
?ksvm

## Evaluating model performance ----
# predictions on testing dataset
forest_predictions <- predict(forest_classifier, forest_test)

table(forest_predictions, forest_test$size_category)
agreement <- forest_predictions == forest_test$size_category
table(agreement)
prop.table(table(agreement))

## Improving model performance ----
#forest_classifier_rbf <- ksvm(size_category ~ ., data = forest_train, kernel = "rbfdot")
#forest_predictions_rbf <- predict(forest_classifier_rbf, forest_test)
#agreement_rbf <- forest_predictions_rbf == forest_test$size_category
#table(agreement_rbf)
#prop.table(table(agreement_rbf))

########################## END #############################################

File: SVM.py
################################### problem1 ###############################
import pandas as pd
import numpy as np
from sklearn import preprocessing

salary_train = pd.read_csv("E:/DATA SCIENCE ASSIGNMENT/Class And Assignment Dataset/Asss/Black  Box Tech (Support Vector Machine)/SalaryData_Train.csv")
salary_test = pd.read_csv("E:/DATA SCIENCE ASSIGNMENT/Class And Assignment Dataset/Asss/Black  Box Tech (Support Vector Machine)/SalaryData_Test.csv")
salary_train.describe()
salary_test.describe()
salary_test.columns
salary_test.dtypes
#EDA
#Selecting required column
salary_train = salary_train.drop(salary_train.columns[[4,6,7,8,12]], axis = 1)
salary_test = salary_test.drop(salary_test.columns[[4,6,7,8,12]], axis = 1)

#Label Encoding for train and test data
label_encoder = preprocessing.LabelEncoder()
salary_train['workclass'] = label_encoder.fit_transform(salary_train['workclass'])
salary_train['education'] = label_encoder.fit_transform(salary_train['education'])
salary_train['occupation'] = label_encoder.fit_transform(salary_train['occupation'])
salary_train['Salary'] = label_encoder.fit_transform(salary_train['Salary'])

label_encoder = preprocessing.LabelEncoder()
salary_test['workclass'] = label_encoder.fit_transform(salary_test['workclass'])
salary_test['education'] = label_encoder.fit_transform(salary_test['education'])
salary_test['occupation'] = label_encoder.fit_transform(salary_test['occupation'])
salary_test['Salary'] = label_encoder.fit_transform(salary_test['Salary'])

from sklearn.svm import SVC
from sklearn.model_selection import train_test_split

train_X = salary_train.iloc[:, 0:8]
train_y = salary_train.iloc[:, 8]
test_X  = salary_test.iloc[:, 0:8]
test_y  = salary_test.iloc[:, 8]


# kernel = linear
help(SVC)
model_linear = SVC(kernel = "linear")
model_linear.fit(train_X, train_y)
pred_test_linear = model_linear.predict(test_X)

np.mean(pred_test_linear == test_y)

# kernel = rbf
model_rbf = SVC(kernel = "rbf")
model_rbf.fit(train_X, train_y)
pred_test_rbf = model_rbf.predict(test_X)

np.mean(pred_test_rbf==test_y)


################################## problem2 ######################################
import pandas as pd
import numpy as np
from sklearn import preprocessing

forestfires_data = pd.read_csv("E:/DATA SCIENCE ASSIGNMENT/Class And Assignment Dataset/Asss/Black  Box Tech (Support Vector Machine)/forestfires.csv")
forestfires_data.describe()
forestfires_data.columns

#EDA
#Label Encoding
label_encoder = preprocessing.LabelEncoder()
forestfires_data['month'] = label_encoder.fit_transform(forestfires_data['month'])
forestfires_data['day'] = label_encoder.fit_transform(forestfires_data['day'])
forestfires_data['size_category'] = label_encoder.fit_transform(forestfires_data['size_category'])

from sklearn.svm import SVC
from sklearn.model_selection import train_test_split

train,test = train_test_split(forestfires_data, test_size = 0.20)

train_X = train.iloc[:, 0:30]
train_y = train.iloc[:, 30]
test_X  = test.iloc[:, 0:30]
test_y  = test.iloc[:, 30]


# kernel = linear
help(SVC)
model_linear = SVC(kernel = "linear")
model_linear.fit(train_X, train_y)
pred_test_linear = model_linear.predict(test_X)

np.mean(pred_test_linear == test_y)

# kernel = rbf
model_rbf = SVC(kernel = "rbf")
model_rbf.fit(train_X, train_y)
pred_test_rbf = model_rbf.predict(test_X)

np.mean(pred_test_rbf==test_y)

